### PR TITLE
refactor(core): minor clean up of the history sampling code

### PR DESCRIPTION
**Pull request by kptkin**
Description
-----------

What does the PR do? Include a concise description of the PR contents.

This PR just moves some code around and adds a little bit of documentation. 
Unfortunately, i wasn't able to find the paper on which this version of the reservoir sampling was implemented. 

- [x] I updated CHANGELOG.md, or it's not applicable

Testing
-------
How was this PR tested?

### CODE REVIEW

File: core/internal/sampler/reservoir_samper.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	ratio := float64(rs.k) / float64(seen)
+	q := math.Min(1, ratio+gamma+math.Sqrt(math.Pow(gamma, 2)+2*gamma*ratio))
+
+	// generate a random priority for the current item
+	x := rand.Float64()
+
+	// add the item to the reservoir if its priority is less than q
+	if x < q {
+		item := &Item[T]{value: value, priority: x, index: rs.seen}
+		heap.Push(&rs.pq, item)
```
[the remark this answers is not in the document]
Comment (kptkin) [reply core/internal/sampler/reservoir_samper.go:53]: Could be, I implemented an algorithm that was in a paper that I can’t find anymore. We can revisit it and replace it with a different algorithm going forward. I will leave a TODO here.

File: core/internal/sampler/reservoir_samper.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	return &ReservoirSampler[T]{
+		pq:    make(PriorityQueue[T], 0),
+		k:     k,
+		delta: delta,
+		seen:  0,
+	}
+}
+
+// Add adds a new item to the reservoir with the given value.
+// The item is added to the reservoir with probability q.
```
[the remark this answers is not in the document]
Comment (kptkin) [reply core/internal/sampler/reservoir_samper.go:38]: Sure, I’ll remove it

File: core/internal/sampler/priority_queue.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	// index is the order in which the item was added
+	index int
+}
+
+// String returns the string representation of the item's value.
+func (item Item[T]) String() string {
+	return fmt.Sprintf("%v", item.value)
+}
+
+// PriorityQueue represents a queue of Items, ordered by their priority.
```
[the remark this answers is not in the document]
Comment (kptkin) [reply core/internal/sampler/priority_queue.go:25]: I’m was debating if I should refactor it as I just moved the code around, but I could do it.

File: core/internal/sampler/reservoir_samper.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (timoffex) [reply core/internal/sampler/reservoir_samper.go:53]: Sounds good, can do it later! Can you make a Jira ticket for it? Seems important to get right for users who have very long runs and log a large number of metrics.

[the remark this answers is not in the document]
Comment (kptkin) [reply core/internal/sampler/reservoir_samper.go:53]: agreed! will do!

[the remark this answers is not in the document]
Comment (kptkin) [reply core/internal/sampler/reservoir_samper.go]: https://wandb.atlassian.net/browse/WB-18332

**Review by timoffex — APPROVED** [review]

**Pull request merged**